optaplanner-examples/src/main/java/org/optaplanner/examples/meetingscheduling/optional/score/MeetingSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,224 @@
+        return constraintFactory.fromUnfiltered(MeetingAssignment.class)
+                .filter(meetingAssignment -> meetingAssignment.getRequiredCapacity() > meetingAssignment.getRoomCapacity())
+                .penalizeConfigurable("Required room capacity",
+                        meetingAssignment -> meetingAssignment.getRequiredCapacity() - meetingAssignment.getRoomCapacity());
+    }
+
+    protected Constraint startAndEndOnSameDay(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUnfiltered(MeetingAssignment.class)
+                .filter(meetingAssignment -> meetingAssignment.getStartingTimeGrain() != null)
+                .join(TimeGrain.class,
+                        equal(MeetingAssignment::getLastTimeGrainIndex, TimeGrain::getGrainIndex),
+                        filtering((meetingAssignment,
+                                timeGrain) -> meetingAssignment.getStartingTimeGrain().getDay() != timeGrain.getDay()))
+                .penalizeConfigurable("Start and end on same day");
+    }
+
+    // ************************************************************************
+    // Medium constraints
+    // ************************************************************************
+
+    protected Constraint requiredAndPreferredAttendanceConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(RequiredAttendance.class)
+                .join(PreferredAttendance.class,
+                        equal(RequiredAttendance::getPerson, PreferredAttendance::getPerson))
+                .join(MeetingAssignment.class,
+                        filtering((requiredAttendance, preferredAttendance,
+                                leftAssignment) -> requiredAttendance.getMeeting() == leftAssignment.getMeeting()),
+                        filtering((requiredAttendance, preferredAttendance,
+                                leftAssignment) -> leftAssignment.getStartingTimeGrain() != null))
+                .join(MeetingAssignment.class,
+                        filtering((requiredAttendance, preferredAttendance, leftAssignment,
+                                rightAssignment) -> preferredAttendance.getMeeting() == rightAssignment.getMeeting()),
+                        filtering((requiredAttendance, preferredAttendance, leftAssignment,
+                                rightAssignment) -> rightAssignment.calculateOverlap(leftAssignment) > 0))
+                .penalizeConfigurable("Required and preferred attendance conflict");
+    }
+
+    protected Constraint preferredAttendanceConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(PreferredAttendance.class)
+                .join(PreferredAttendance.class,
+                        equal(PreferredAttendance::getPerson, PreferredAttendance::getPerson),
+                        lessThan(PreferredAttendance::getId, PreferredAttendance::getId))
```

Comment:
`fromUniquePair(...)`?

Also, when operating on objects of the same type, `equal(PreferredAttendance::getPerson)` is syntactic sugar for `equal(PreferredAttendance::getPerson, PreferredAttendance::getPerson)`.

I believe the same is true for `lessThan()`.

Reply:
Ah, good to know, wasn't aware of the `equal(PreferredAttendance::getPerson)` syntactic sugar. As for `fromUniquePair`, I actually decided against using it since doing the `join` manually reads better for me. I don't have an issue to changing it, but in general is syntactic sugar preferred?